File: 2018/aeds/pilha/java/Pilha.java
public class Pilha {
   private int tamanho;
   private int[] vetor;
   private int topo = -1;

   public Pilha(int tamanho) {
      this.tamanho = tamanho;
      vetor = new int[tamanho];
   }

   public boolean vazia() {
      return topo == -1;
   }

   public boolean cheia() {
      return topo == tamanho-1;
   }

   public void push(int e) {
      if (!cheia())
         vetor[++topo] = e;
      else
         throw new IllegalStateException("Pilha Cheia");
   }

   public int pop() {
      if (!vazia())
         return vetor[topo--];
      else
         throw new IllegalStateException("Pilha vazia");
   }

   public static void main(String[] args) {
      System.out.println("Criando uma pilha de 5 elementos");
      Pilha p = new Pilha(5);

      try {
         int i = 1;
         while (true) {
            System.out.printf("Empilhando %d\n", i);
            p.push(i++);
         }
      } catch (Exception e) {
         System.out.printf("Ops! %s\n", e.getMessage());
      }

      System.out.println("Agora, desempilhando...");
      try {
         while (true)
            System.out.printf("Desempilhou %d\n", p.pop());
      } catch (Exception e) {
         System.out.printf("Ops! %s\n", e.getMessage());
      }
   }
}

File: 2018/aeds/pilha/python/pilha.py
# -*- coding: utf-8 -*-

# Implementação "agnóstica" de pilha em Python

class Pilha:
    def __init__(self, tamanho):
        self.tamanho = tamanho
        self.vetor = [None] * tamanho
        self.topo = -1
    
    def vazia(self):
        return self.topo == -1
    
    def cheia(self):
        return self.topo == self.tamanho-1
    
    def push(self, e):
        if not self.cheia():
            self.topo += 1
            self.vetor[self.topo] = e
        else:
            raise Exception('pilha cheia')
    
    def pop(self):
        if not self.vazia():
            self.topo -= 1
            return self.vetor[self.topo+1]
        else:
            raise Exception('Pilha vazia')

if __name__ == "__main__":
    print "Testando a pilha"
    print "Criando uma pilha de 5 elementos"
    p = Pilha(5)
    try:
        for i in range(1,10):
            print 'Tentando empilhar %d' % i,
            p.push(i)
            print 'empilhou'
    except Exception as e:
        print 'oops! ', e
    
    print "Desempilhando..."
    try:
        while True:
            print "Desempilhou ", p.pop()
    except Exception as e:
        print "OOps! ", e


